bagua/torch_api/distributed.py

```diff
@@ -93,23 +98,100 @@ def produces_sparse_gradient(module):
 
         return parameters
 
+    def _bagua_broadcast_optimizer_state(self, optimizer):
+        # L-BFGS cannot be easily supported without serializing
+        # the entire state_dict, as its structure is deeply nested and contains
+        # None type parameter values.
+        if isinstance(optimizer, torch.optim.LBFGS):
+            raise ValueError("cannot broadcast torch.optim.LBFGS state")
+        optimizer_state_dict = optimizer.state_dict()
+
+        # Initialize newly created optimizers.
+        if len(optimizer_state_dict["state"]) == 0:
+            for group in optimizer.param_groups:
+                for p in group["params"]:
+                    if p.requires_grad and id(p) not in optimizer_state_dict["state"]:
+                        p.grad = p.data.new(p.size()).zero_()
+                        if isinstance(optimizer, torch.optim.SparseAdam):
+                            p.grad = p.grad.to_sparse()
+            optimizer.step()
```

Comment:
We must call `optimizer.step` to get the state dict?

Reply:
Thanks for your review. It seems some states will be initialized in the first 'optimizer.step' call like shown in https://github.com/pytorch/pytorch/blob/1b1f1e36b42fd9042ba7b7c37f4a717b6631edff/torch/optim/sgd.py#L105 and https://github.com/pytorch/pytorch/blob/1b1f1e36b42fd9042ba7b7c37f4a717b6631edff/torch/optim/adam.py#L86., so this should be necessary.